Add a separate target kinematics_cache_helper

hongkai-dai opened the pull request:
I find I need `KinematicsCacheHelper` in trajectory optimization. In this PR, I did three things
1. Move `KinematicsCacheHelper` to a separate file, and build it to a standalone target.
2. Instantiate `KinematicsCacheHelper` with `AutoDiffXd`. This is because in trajectory optimization, we need to compute the gradient of Jᵀλ w.r.t q. Thus we need to compute the Jacobian `J ` using `AutoDiffXd`.
3. Add another class `KinematicsCacheWithVHelper`, which compute a `KinematicsCache` with both position `q` and velocity `v`. This is used when we compute the gradient of the dynamics, for example, the gradient of the mass matrix or the Coriolis force.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7978)
<!-- Reviewable:end -->

hongkai-dai commented:
+@sammy-tri for feature review please.

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-:-L4j8v7IPY9Wu39XU9wk:bjquaum)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/kinematics_cache_helper.h, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4ksKO69qlFv4PSdUyV:-L4lH7AgGg8Gwp4Pax8r:b4a3n2x) ([raw file](https://github.com/robotlocomotion/drake/blob/f450be3cd699827f408dfd20d35542159b82fcdf/multibody/kinematics_cache_helper.h#L34)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'd prefer we be consistent about defining code in headers vs. cc files, at least when there are two classes this similar in the same header.  I don't have a strong preference for which way it goes in this case.
</blockquote></details>

Done. I put it in cc file.

---

*[multibody/kinematics_cache_helper.h, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4ktKJ4D299lfxXSUaP:-L4lIWOVt_sMwXxIlk22:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f450be3cd699827f408dfd20d35542159b82fcdf/multibody/kinematics_cache_helper.h#L34)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

We should document that `tree` is aliased and needs to live for the lifetime of the object.
</blockquote></details>

Done.

---

*[multibody/kinematics_cache_helper.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4ksvEbQt2VklRgvBP1:-L4lIjiUbCXirzHcDiVy:b-xlfeoh) ([raw file](https://github.com/robotlocomotion/drake/blob/f450be3cd699827f408dfd20d35542159b82fcdf/multibody/kinematics_cache_helper.h#L37)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I don't see where `last_v_` is ever set to a non-zero value.  This seems strange enough to me that it could use a comment if it's deliberate.
</blockquote></details>

Done. Doh! Forgot to include the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
One comment left.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/kinematics_cache_helper.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4ksvEbQt2VklRgvBP1:-L4lKt5nM7F0QAw8gxTm:b-f06ar) ([raw file](https://github.com/robotlocomotion/drake/blob/f450be3cd699827f408dfd20d35542159b82fcdf/multibody/kinematics_cache_helper.h#L37)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. Doh! Forgot to include the code.
</blockquote></details>

Thanks!  I think it's probably worth documenting on the `q` only version that it will use whatever velocity was last passed into the version which takes both `q` and `z`?  (as opposed to other assumptions like that it might default to zero even if the two-argument version had been called before?)  (actually I'm not totally clear on the use for the one-argument version vs. the other version of `KinematicsCacheHelper`.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-:-L4lLIk1w-IToEd7cw5H:b-qdirw9)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 5 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/kinematics_cache_helper.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4ksvEbQt2VklRgvBP1:-L4lO6yxRCe3qSuGeiIc:b-c9d57z) ([raw file](https://github.com/robotlocomotion/drake/blob/f450be3cd699827f408dfd20d35542159b82fcdf/multibody/kinematics_cache_helper.h#L37)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Thanks!  I think it's probably worth documenting on the `q` only version that it will use whatever velocity was last passed into the version which takes both `q` and `z`?  (as opposed to other assumptions like that it might default to zero even if the two-argument version had been called before?)  (actually I'm not totally clear on the use for the one-argument version vs. the other version of `KinematicsCacheHelper`.)
</blockquote></details>

Done. I added some documentation here, do you think it is clear?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a request for clarification.

---

Reviewed 5 of 7 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/kinematics_cache_helper.h, line 30 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4lY7RycguzbNw47iUU:-L4lY7RycguzbNw47iUV:b-bcpoc7) ([raw file](https://github.com/robotlocomotion/drake/blob/849c5a26147c38497bd3e6b89d1b1f69534b302b/multibody/kinematics_cache_helper.h#L30)):*
> ```objc
>   VectorX<Scalar> last_q_;
>   const RigidBodyTree<double>* last_tree_;
>   KinematicsCache<Scalar> kinsol_;
> ```

BTW The name "kinsol" seems obscure.  Is it documented somewhere?

---

*[multibody/kinematics_cache_helper.h, line 77 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4lYI5POhiSTLuHtfH_:-L4lYI5POhiSTLuHtfHa:b-42bu02) ([raw file](https://github.com/robotlocomotion/drake/blob/849c5a26147c38497bd3e6b89d1b1f69534b302b/multibody/kinematics_cache_helper.h#L77)):*
> ```objc
>   VectorX<Scalar> last_q_;
>   VectorX<Scalar> last_v_;
>   KinematicsCache<Scalar> kinsol_;
> ```

BTW As above:  The name "kinsol" seems obscure.  Is it documented somewhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-:-L4lYMlPrIWPqYgbi32d:bkzzwqk)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[multibody/kinematics_cache_helper.h, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4ksvEbQt2VklRgvBP1:-L4lceCtmi1MAJUxDM6C:buvhge2) ([raw file](https://github.com/robotlocomotion/drake/blob/f450be3cd699827f408dfd20d35542159b82fcdf/multibody/kinematics_cache_helper.h#L37)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Thanks!
</blockquote></details>

You are right, it is very confusing to update `q` only. I remove this function in this PR. 

---

*[multibody/kinematics_cache_helper.h, line 30 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4lY7RycguzbNw47iUU:-L4lcA-Uq3PNU_YVj5xH:b18h5u0) ([raw file](https://github.com/robotlocomotion/drake/blob/849c5a26147c38497bd3e6b89d1b1f69534b302b/multibody/kinematics_cache_helper.h#L30)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW The name "kinsol" seems obscure.  Is it documented somewhere?
</blockquote></details>

Done. It was inherited from the old code https://github.com/RobotLocomotion/drake/blob/master/multibody/constraint_wrappers.cc#L10, I changed the name to `kinematics_cache`.

---

*[multibody/kinematics_cache_helper.h, line 53 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4lUpcvZHpkfL_PD6Jp:-L4lcnArFseRB7NnqM3F:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/849c5a26147c38497bd3e6b89d1b1f69534b302b/multibody/kinematics_cache_helper.h#L53)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW typo (missing 'e' in 'Cache')
</blockquote></details>

Done.

---

*[multibody/kinematics_cache_helper.h, line 77 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-L4lYI5POhiSTLuHtfH_:-L4lco5NsmEk1chDwM_d:b77l429) ([raw file](https://github.com/robotlocomotion/drake/blob/849c5a26147c38497bd3e6b89d1b1f69534b302b/multibody/kinematics_cache_helper.h#L77)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW As above:  The name "kinsol" seems obscure.  Is it documented somewhere?
</blockquote></details>

Done. Changed it to `kinematics_cache_`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@drake-jenkins-bot linux-xenial-gcc-cmake-experimental-matlab plsease.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-:-L4lrZ-eG3yWYUkLnktK:b-y0rz0x)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@drake-jenkins-bot linux-xenial-gcc-cmake-experimental-matlab please.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-:-L4lrtAwLOgFwIlcufy3:bzcj60u)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 1 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7978#-:-L4m-4AI7IkUV_y1UUCS:b-l70wzf)*
<!-- Sent from Reviewable.io -->
